### DISCUSSION TITLE
Move away from auto importing every instrument in instruments/__init__.py

### DISCUSSION THREAD

**Discussion by mcdo0486** [pymeasure/pymeasure#801] [Ideas]
Currently over 90 instruments are auto imported in [`pymeasure/instruments/__init__.py`](https://github.com/pymeasure/pymeasure/blob/2cfff5f629148e5da07810f3b4384b6de5ad7557/pymeasure/instruments/__init__.py)

That means if you import an instrument directly, such as `from pymeasure.instruments.hp.hp3437A import HP3437A`, the transversal over pymeasure/instruments will import every other instrument as well.

Some instruments, like [ dsp7265](https://github.com/pymeasure/pymeasure/blob/2cfff5f629148e5da07810f3b4384b6de5ad7557/pymeasure/instruments/signalrecovery/dsp7265.py#L124), create large in-memory lists for value validation. To be clear, I don't think the issue is in the details of how each instrument is implemented, it is the auto importing of every instrument. 

Using [memray](https://github.com/bloomberg/memray), the instrument auto imports take up about ~10 MB of the ~100MB memory footprint of pymeasure.

As more instruments are added, and more adapter dependencies are added for those instruments, the memory footprint will only continue to grow. I've been looking through some optimizations to make pymeasure launch faster and be more responsive on slower computers with less available memory. I removed all auto imports in [this branch](https://github.com/mcdo0486/pymeasure/tree/optimize/remove_inst_auto_imports), only a handful of instruments needed updates to, for instance, import the `discreteTruncate` validator correctly.

This would break backwards compatibility for procedures that import instruments from pymeasure.instruments directly.

**Comment by BenediktBurger** [comment]
Thanks for raising this question.

I always wondered, what the cost of loading all instruments is. Thanks for giving that number.

At a first glance, I personally think, that we could drop all imports in `instrument.__init__.py`.
Loading all instruments of one manufacturer should be OK, I guess.

**Reply by mcdo0486** [reply]
Yea, I also was thinking auto imports at the vendor level is ok. 

We can see what others think about the instrument import change, and vendor level imports,  then I can submit my PR.

**Reply by BenediktBurger** [reply]
Another reason, why this is good: importing an instrument won't try to load nidaqmx for the virtual bench etc.

I just looked at your example branch:
- I agree with removing the auto loading of instruments.
- I disagree with removing auto loading the adapters, as they are only a few (and won't become many more) and it is more practical to import them that easily. You don't save any space with not loading the `VISAAdapter`, as `Instrument` loads that anyways.
- I'm unsure about not loading the validators/error into the instruments namespace.

**Reply by bilderbuchi** [reply]
> I'm unsure about not loading the validators/error into the instruments namespace.

Those should (gut feeling) have a negligible impact on loading time.

**Reply by mcdo0486** [reply]
There is only one validator that is currently in `pymeasure/instruments/__init__.py`, `discreteTruncate`:
https://github.com/pymeasure/pymeasure/blob/eab361ccfef345a6ffd347868ef68cfa25d21aaf/pymeasure/instruments/__init__.py#L29

For adapters, I think the current list of adapters can be kept in `pymeasure.adapters.__init__.py`. But I can see more adapters being added for instruments  drivers, I've added `minimalmodbus` and `pyusb` to support some instruments and I can imagine newer instruments from smaller companies not implementing VISA. That said, some instrument specific adapters can be kept in the instrument folder, and they can have their own non-required dependencies like `pymeasure.instruments.ni`. Though that wouldn't be as user friendly, not all instrument drivers "just work" after pymeasure install.

I don't think there will be a good time for this switch, so I'll submit that branch as a PR for consideration and further discussion. 

PR: https://github.com/pymeasure/pymeasure/pull/919

**Reply by BenediktBurger** [reply]
Regarding Adapters:
I do not see the necessity for many adapters: visa offers access to raw serial and raw USB.

**Comment by bilderbuchi** [comment]
Memray seems like a useful tool I didn't know, thanks!

In principle the optimisation sounds reasonable. We have to think about how we approach the backwards-compatibility issue -- can we use some `importlib` thing or whatnot to catch people relying on the autoimport (i.e. catch the `ImportError` from "inside", so to say), issue a warning, and import the "old" way on demand only?

Another thing we should take into account: What's in `instruments/__init__` defines a public, discoverable, API, and, crucially, allows frictionless autocompletion for discovery of manufacturers and instruments! In IPython and friends you can do `import pym<tab>.instr<tab>.ke<tab>.<tab>` to quickly find all available Keithley instruments. Does that still work with your approach?

**Reply by mcdo0486** [reply]
> In IPython and friends you can do `import pym<tab>.instr<tab>.ke<tab>.<tab>` to quickly find all available Keithley instruments. Does that still work with your approach?

Yup autocompletion still works without having everything in `__init__.py`, and I agree that is a crucial feature.  If tab completion didn't work then I'd still be typing my first pull request.

**Reply by BenediktBurger** [reply]
I guess it finds modules / subpackeges and allows autocompletion (similar to "pymeasure.instruments", which works without an import).

**Reply by bilderbuchi** [reply]
That's great to hear!

I could not come up with a way to avoid the backwards-incompatible-breaking removal of the imports, I guess we can't avoid that. OTOH, the ImportError should be pretty clear, and we can add a notice to the changelog and maybe documentation.
